New topic: Creating localized exception messages (#14147)
dotnet/docs: docs/standard/exceptions/creating-localized-exception-messages.md

---

> title: How to: create custom exceptions and localized exception messages

Author: Can you take a look on https://github.com/dotnet/docs/pull/14141 ?
@rpetrusha

> title: How to: create custom exceptions and localized exception messages

Author: I've changed `custom` with `user-defined` & `and` with `with`. Left `create` lower case for now until you check #14141 .

> title: How to: create custom exceptions and localized exception messages

Reviewer: Yes, leaving `create` lowercase conforms to our current style, @Youssef1313.

description: The article demonstrates how to create custom exceptions and localized exception messages
author: Youssef1313
ms.date: 09/02/2019
---
# How to: create custom exceptions and localized exception messages

> Written by [Youssef Victor](https://github.com/Youssef1313).

Author: I thought it was okay to include it when I saw [Best practices for writing unit tests](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices)

## Create custom exceptions
The .NET Framework contains many different exceptions that are likely to meet your needs. However, in some cases when none of them meets your needs, you can create your own custom exceptions.
Let's assume you want to create a `StudentNotFoundException` that contains a `StudentName` property.
To create custom exceptions, Follow these steps:

- Create a serializable class that inherits from `Exception`. The class name is preferred to end in "Exception".

> ```csharp

Reviewer: The code here and throughout needs to be indented so that it aligns with the bullets. Here is what it looks like now:

![image](https://user-images.githubusercontent.com/10886961/64463322-79f26780-d0b8-11e9-96ef-7728aab57e02.png)

Author: I'm not sure if it will render correctly now ?

[Serializable]
public class StudentNotFoundException : Exception
{
    
}
```

- Add the default constructors.

```csharp
[Serializable]
public class StudentNotFoundException : Exception
{
    public StudentNotFoundException()
    {
    }

    public StudentNotFoundException(string message)
        : base(message)
    {
    }

    public StudentNotFoundException(string message, Exception inner)
        : base(message, inner)
    {
    }
}
```

> - Add properties and/or more constructors depending on your needs.

Reviewer: ```suggestion
- Add properties and more constructors, depending on your needs:
```

```csharp
[Serializable]
public class StudentNotFoundException : Exception
{
    public string StudentName { get; }

    public StudentNotFoundException()
    {
    }

    public StudentNotFoundException(string message)
        : base(message)
    {
    }

    public StudentNotFoundException(string message, Exception inner)
        : base(message, inner)
    {
    }
	
    public StudentNotFoundException(string message, string studentName)
        : base(message)
    {
        StudentName = studentName;
    }
}
```

## Create localized exception messages

> You have created our custom exception and you can throw it anywhere in code like the following:

Reviewer: ```suggestion
You have created a custom exception, and you can throw it anywhere with code like the following:
```

```csharp
throw new StudentNotFoundException("The student cannot be found.", "John");
```

> The previous line looks good, but its only problem is that `The student cannot be found.` is just a constant string, you want to have different messages depending on user culture.

Reviewer: ```suggestion
The problem with the previous line is that "The student cannot be found." is just a constant string. In a localized application, you want to have different messages depending on user culture.
```

> [Satellite Assemblies](https://docs.microsoft.com/dotnet/framework/resources/creating-satellite-assemblies-for-desktop-apps) is the good way to do that.

Reviewer: ```suggestion
[Satellite Assemblies](https://docs.microsoft.com/dotnet/framework/resources/creating-satellite-assemblies-for-desktop-apps) are a good way to do that.
```

> A satellite assembly is a `.dll` that contains resources for a specific language. When you ask for a specific resources at runtime, the CLR finds that resource depending on user culture. If no satellite assembly is found for that culture, the default resources file is used.

Reviewer: ```suggestion
A satellite assembly is a .dll that contains resources for a specific language. When you ask for a specific resources at runtime, the CLR finds that resource depending on user culture. If no satellite assembly is found for that culture, the resources of the default culture are used.
```

> To start creating the localized exception messages:

Reviewer: ```suggestion
To create the localized exception messages:
```

> - Create a new folder to hold the resource files, let's name the folder `Resources`.

Reviewer: ```suggestion
- Create a new folder named *Resources* to hold the resource files.
```

> - Add a new resource file to it. To do that in Visual Studio, Right click the folder in solution explorer -> Add -> New Item -> Choose `Resources File` -> Name it `ExceptionMessages.resx`. (This is the default resources file)

Reviewer: ```suggestion
- Add a new resource file to it. To do that in Visual Studio, right-click the folder in **Solution Explorer**, and select **Add** -> **New Item** -> **Resources File**. Name the file *ExceptionMessages.resx*. This is the default resources file.
```

> - Add a Name Value pair for your exception messages like the following image shows:

Reviewer: ```suggestion
- Add a name/value pair for your exception message, like the following image shows:
```

> ![image](https://user-images.githubusercontent.com/31348972/64120930-ca5a8400-cd9d-11e9-984a-789c5f582513.png)

Reviewer: Images should be included with this PR in a *media* subdirectory of the directory that contains the article. It would be good to reshoot the image to get rid of the blue highlighting, which renders the text nearly invisible.

> - Add a new resource file for French. Name it `ExceptionMessages.fr-FR.resx`.

Reviewer: ```suggestion
- Add a new resource file for French. Name it *ExceptionMessages.fr-FR.resx*.
```

> - Add a Name Value pair of the exception messages again but with French value:

Reviewer: ```suggestion
- Add a name/value pair for the exception message again, but with a French value:
```

> ![image](https://user-images.githubusercontent.com/31348972/64120995-ee1dca00-cd9d-11e9-9728-07b971bb6409.png)

Reviewer: Again, this should be in the *media* directory, not in a Gist.

> - If you build the project and go `bin/Debug`, you will find a folder named `fr-FR` containing a `.dll` file, this is our satellite assembly.

Reviewer: ```suggestion
- If you build the project and go to the *bin/Debug* folder, you'll find a folder named *fr-FR* containing a .dll file; this is the satellite assembly.
```

> - When you want to  throw the exception, you do the following:

Reviewer: ```suggestion
- You throw the exception with code like the following:
```

```csharp
var resourceManager = new ResourceManager("FULLY_QIALIFIED_NAME_OF_RESOURCE_FILE", Assembly.GetExecutingAssembly());
throw new StudentNotFoundException(resourceManager.GetString("StudentNotFound"), "John")
```

> [!NOTE]

> > If our project name is `TestProject` and our resource file is called `ExceptionMessages.resx` and is inside a folder called `Resources` in the project, the fully qualified name will be `TestProject.Resources.ExceptionMessages`.

Reviewer: ```suggestion
> If our project name is `TestProject` and our resource file is *ExceptionMessages.resx* and is in a folder named *Resources* in the project, the fully qualified name of the resource file is `TestProject.Resources.ExceptionMessages`.
```
You should also add a **See also** section that includes cross-references to the "How to; Create user-defined exceptions" article as well as Creating Satellite Assemblies for Desktop Apps.

---
title: "How to: create user-defined exceptions with localized exception messages"
description: "Learn how to create user-defined exceptions with localized exception messages"
author: Youssef1313
ms.author: ronpet
ms.date: 09/02/2019
---
# How to: create user-defined exceptions with localized exception messages

> 

Author: I can't really find a good intro :(

> 

Author: I've added a small intro. not sure if it's good.

    ```csharp
    [Serializable]
    public class StudentNotFoundException : Exception
    {
    
    }
    ```

- Add the default constructors:

    ```csharp
    [Serializable]
    public class StudentNotFoundException : Exception
    {
        public StudentNotFoundException()
        {
        }

        public StudentNotFoundException(string message)
            : base(message)
        {
        }

        public StudentNotFoundException(string message, Exception inner)
            : base(message, inner)
        {
        }
    }
    ```

- Add properties and more constructors, depending on your needs:

    ```csharp
    [Serializable]
    public class StudentNotFoundException : Exception
    {
        public string StudentName { get; }

        public StudentNotFoundException()
        {
        }

        public StudentNotFoundException(string message)
            : base(message)
        {
        }

        public StudentNotFoundException(string message, Exception inner)
            : base(message, inner)
        {
        }
	
        public StudentNotFoundException(string message, string studentName)

>             : base(message)

Author: For the example used in the article it won't make a difference. I think that in most cases, the user will want the custom logic to be called. I'll replace `base` with `this`. I'll also add links to the `base` keyword page and `this` keyword page in the **see also** section.

A satellite assembly is a .dll that contains resources for a specific language. When you ask for a specific resources at runtime, the CLR finds that resource depending on user culture. If no satellite assembly is found for that culture, the resources of the default culture are used.

To create the localized exception messages:
- Create a new folder named *Resources* to hold the resource files.
- Add a new resource file to it. To do that in Visual Studio, right-click the folder in **Solution Explorer**, and select **Add** -> **New Item** -> **Resources File**. Name the file *ExceptionMessages.resx*. This is the default resources file.
- Add a name/value pair for your exception message, like the following image shows:

![Add resources to the default culture](media/add-resources-to-default-culture.jpg)

- Add a new resource file for French. Name it *ExceptionMessages.fr-FR.resx*.
- Add a name/value pair for the exception message again, but with a French value:

![Add resources to the fr-FR culture](media/add-resources-to-fr-culture.jpg)

> - If you build the project and go to the *bin/Debug* folder, you'll find a folder named *fr-FR* containing a .dll file; this is the satellite assembly.

Reviewer 2: One can build the project in Release or to the different folder:
```suggestion
- After you build the project, the build output folder should contain the *fr-FR* folder with a *.dll* file, which is the satellite assembly.
```

- You throw the exception with code like the following:

    ```csharp
    var resourceManager = new ResourceManager("FULLY_QIALIFIED_NAME_OF_RESOURCE_FILE", Assembly.GetExecutingAssembly());
    throw new StudentNotFoundException(resourceManager.GetString("StudentNotFound"), "John");
    ```

> > [!NOTE]

Reviewer 2: The note belongs to the list item, so indent it as well:
```suggestion
  > [!NOTE]
```

> > If our project name is `TestProject` and our resource file is *ExceptionMessages.resx* and is in a folder named *Resources* in the project, the fully qualified name of the resource file is `TestProject.Resources.ExceptionMessages`.

Reviewer 2: The docs don't use "I", "we", "us", "our"...
```suggestion
> If the project name is `TestProject` and the resource file *ExceptionMessages.resx* resides in the *Resources* folder of the project, the fully qualified name of the resource file is `TestProject.Resources.ExceptionMessages`.
```

## See also

> [How to create user-defined exceptions](how-to-create-user-defined-exceptions.md)

Reviewer 2: **See also** sections are formatted as lists:
```suggestion
- [How to create user-defined exceptions](how-to-create-user-defined-exceptions.md)
```

> 

Reviewer 2: No blank lines between list items in the **See also** sections:
```suggestion
```

> [Creating Satellite Assemblies for Desktop Apps](../../framework/resources/creating-satellite-assemblies-for-desktop-apps.md)

Reviewer 2: Also, the file typically ends with one blank line:
```suggestion
- [Creating Satellite Assemblies for Desktop Apps](../../framework/resources/creating-satellite-assemblies-for-desktop-apps.md)

```

- Create a new folder named *Resources* to hold the resource files.
- Add a new resource file to it. To do that in Visual Studio, right-click the folder in **Solution Explorer**, and select **Add** -> **New Item** -> **Resources File**. Name the file *ExceptionMessages.resx*. This is the default resources file.
- Add a name/value pair for your exception message, like the following image shows:

![Add resources to the default culture](media/add-resources-to-default-culture.jpg)

- Add a new resource file for French. Name it *ExceptionMessages.fr-FR.resx*.
- Add a name/value pair for the exception message again, but with a French value:

![Add resources to the fr-FR culture](media/add-resources-to-fr-culture.jpg)

- After you build the project, the build output folder should contain the *fr-FR* folder with a *.dll* file, which is the satellite assembly.

- You throw the exception with code like the following:

    ```csharp
    var resourceManager = new ResourceManager("FULLY_QIALIFIED_NAME_OF_RESOURCE_FILE", Assembly.GetExecutingAssembly());
    throw new StudentNotFoundException(resourceManager.GetString("StudentNotFound"), "John");
    ```

  > [!NOTE]

> > If the project name is `TestProject` and the resource file *ExceptionMessages.resx* resides in the *Resources* folder of the project, the fully qualified name of the resource file is `TestProject.Resources.ExceptionMessages`.

Reviewer 2: Oh, I forgot to indent the second line of the note block:
```suggestion
  > If the project name is `TestProject` and the resource file *ExceptionMessages.resx* resides in the *Resources* folder of the project, the fully qualified name of the resource file is `TestProject.Resources.ExceptionMessages`.
```
